### PR TITLE
Cap on stableToken CP-DOTO tank size

### LINKED ISSUE #5849: [Protocol Economics] Cap on cUSD CP-DOTO tank size
### Expected Behavior

As a safeguard against oracle manipulation or other issues related to an incorrect cUSD/CELO exchange rate, there should be a cap on the cUSD tank size. The rationale is that if other safeguards fail, the impact is limited

### Current Behavior

There is no cap.

### Definition of Done

- We have a design doc detailing why the cap is necessary and how we come up with a value for it 
- Cap is implemented and live on mainnet.

### PR DESCRIPTION
**Pull request by martinvol**
### Description

Adds a maximum size to the stableToken bucket of CP-DOTO.

The cap is a function of the current supply of stableToken, and it's calculated as follows:

`maxStableBucketSize = stableTokenSupply* stableBucketFractionCap`. If the calculated cap after a rebalancing is bigger than the cap, then the goldBucket shall be downsized accordingly.

For that to work, the `stableTokenSupply` needs to be big enough to generate a usable stableTokenBucket size, and it is unusable if a token has a zero supply (because the bucket will be always capped at zero). `minSupplyForStableBucketCap` is introduced for that, and the bucket will always be capped such that it's always bigger than `minSupplyForStableBucketCap* stableBucketFractionCap `

### Added governable variables:
* minSupplyForStableBucketCap
* stableBucketFractionCap

### TODO:

- [x] Validate approach
- [x] Verify it doesn't break proxy storage
- [x] Optimize gas usage
- [x] More testing
- [x] Replace hard-coded constants on test

### Tested

Unit tested added.

### Related issues

- Closes #5849

### Backwards compatibility

_Brief explanation of why these changes are/are not backwards compatible._

### CODE REVIEW

File: packages/protocol/contracts/stability/Exchange.sol [hunk 1]
```diff
@@ -1,6 +1,7 @@
 pragma solidity ^0.5.13;
 
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/math/Math.sol";
```
Comment (martinvol) [inline packages/protocol/contracts/stability/Exchange.sol:4]: will check if I actually need this

File: packages/protocol/contracts/stability/Exchange.sol [hunk 2]
```diff
@@ -64,7 +71,7 @@ contract Exchange is
    * @return The storage, major, minor, and patch version of the contract.
    */
   function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
-    return (1, 1, 1, 0);
+    return (1, 1, 2, 0);
```
Comment (martinvol) [inline packages/protocol/contracts/stability/Exchange.sol:73]: Do I need to do this?

File: packages/protocol/contracts/stability/Exchange.sol [hunk 3]
```diff
@@ -34,8 +35,11 @@ contract Exchange is
   event MinimumReportsSet(uint256 minimumReports);
   event StableTokenSet(address indexed stable);
   event SpreadSet(uint256 spread);
+  event MaxStableBucketFractionSet(uint256 maxStableBucketFraction);
   event ReserveFractionSet(uint256 reserveFraction);
   event BucketsUpdated(uint256 goldBucket, uint256 stableBucket);
+  event StableBucketCapped();
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol]: Seems unused

File: packages/protocol/contracts/stability/Exchange.sol [hunk 4]
```diff
@@ -34,8 +35,11 @@ contract Exchange is
...
+  event minSupplyForStableBucketCapSet(uint256 minSupplyForStableBucketCap);
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol]: Casing

File: packages/protocol/test/stability/exchange.ts [hunk 5]
```diff
@@ -148,7 +148,7 @@ contract('Exchange', (accounts: string[]) => {
     await registry.setAddressFor(CeloContractName.Exchange, exchange.address)
   })
 
-  describe('#initialize()', () => {
+  describe.skip('#initialize()', () => {
```
Comment (nambrot) [inline packages/protocol/test/stability/exchange.ts]: Probably shouldn't skip these

File: packages/protocol/contracts/stability/Exchange.sol (hunk 2 above)
Comment (nambrot) [reply packages/protocol/contracts/stability/Exchange.sol:73]: Yes indeed

File: packages/protocol/contracts/stability/Exchange.sol [hunk 6]
```diff
@@ -54,6 +58,9 @@ contract Exchange is
   uint256 public updateFrequency;
   uint256 public minimumReports;
 
+  uint256 public minSupplyForStableBucketCap;
+  FixidityLib.Fraction public maxStableBucketFraction;
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol]: I'm the worst at naming variables, but I don't know if this feels very descriptive either

File: packages/protocol/contracts/stability/Exchange.sol [hunk 7]
```diff
@@ -283,6 +290,16 @@ contract Exchange is
     emit SpreadSet(newSpread);
   }
 
+  // TODO add docs
+  function setMaxStableBucketFraction(uint256 newMaxStableBucketFraction) public onlyOwner {
+    maxStableBucketFraction = FixidityLib.wrap(newMaxStableBucketFraction);
+    require(
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol:303]: Should we check that it is not 0?

File: packages/protocol/contracts/stability/Exchange.sol [hunk 8]
```diff
@@ -293,6 +310,15 @@ contract Exchange is
     emit ReserveFractionSet(newReserveFraction);
   }
 
+  /**
+    * @notice Allows owner to set the minSupplyForStableBucketCap, the minimal stableToken supply considered for the stableToken cap
+    * @param newMinSupplyForStableBucketCap The new value for the minSupplyForStableBucketCap
+    */
+  function setMinSupplyForStableBucketCap(uint256 newMinSupplyForStableBucketCap) public onlyOwner {
+    minSupplyForStableBucketCap = newMinSupplyForStableBucketCap;
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol:326]: Should we check that it is not 0?

File: packages/protocol/contracts/stability/Exchange.sol [hunk 9]
```diff
@@ -368,9 +394,38 @@ contract Exchange is
     uint256 updatedStableBucket = exchangeRateNumerator.mul(updatedGoldBucket).div(
       exchangeRateDenominator
     );
+
+    uint256 maxStableBUcketSize = getStableBucketTokenCap();
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol]: Typo

File: packages/protocol/contracts/stability/Exchange.sol [hunk 10]
```diff
@@ -368,9 +394,38 @@ contract Exchange is
...
+
+    // check if the token is bigger than the cap
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol]: Token supply?

File: packages/protocol/contracts/stability/Exchange.sol (hunk 6 above)
Comment (nambrot) [reply packages/protocol/contracts/stability/Exchange.sol]: You should probably also add this to the initialization function

File: packages/protocol/contracts/stability/Exchange.sol [hunk 11]
```diff
@@ -368,9 +394,38 @@ contract Exchange is
...
+    }
+
     return (updatedGoldBucket, updatedStableBucket);
   }
 
+  /**
+   * @notice returns the max size the stableToken can be set durin a bucket update. if the supply of stableToken
+   * is smaller than minSupplyForStableBucketCap, the cap is calculated based on minSupplyForStableBucketCap.
+   */
+  function getStableBucketTokenCap() public view returns (uint256) {
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol]: Again not an expert in naming, but this seems a bit off for me. Wouldn't `getStableTokenBucketCap` be more accurate?

File: packages/protocol/contracts/stability/Exchange.sol [hunk 12]
```diff
@@ -368,9 +394,38 @@ contract Exchange is
...
+    uint256 stableTokenSupply = Math.max(IERC20(stable).totalSupply(), minSupplyForStableBucketCap); // assume a miminum of 1M stable tokens supply to start the cap
```
Comment (nambrot) [inline packages/protocol/contracts/stability/Exchange.sol]: Isn't this a somewhat weird situation with let's say `minSupplyForStableBucketCap == 1M` and `maxStableBucketFraction == 0.1`? I.e. the max bucket size is 1M for tokenSupply <1M, but once supply hits 1M, the max is actually 0.1M?

File: packages/protocol/contracts/stability/StableToken.sol [hunk 13]
```diff
@@ -216,7 +216,7 @@ contract StableToken is
    */
   function mint(address to, uint256 value) external updateInflationFactor returns (bool) {
     require(
-      msg.sender == registry.getAddressFor(EXCHANGE_REGISTRY_ID) ||
+      msg.sender == registry.getAddressFor(EXCHANGE_REGISTRY_ID) || // TODO this should be split into multiple CP-DOTO
```
Comment (nambrot) [inline packages/protocol/contracts/stability/StableToken.sol]: I'd probably leave the TODO in a Github Ticket vs. in code

File: packages/protocol/test/stability/exchange.ts [hunk 14]
```diff
@@ -626,6 +626,81 @@ contract('Exchange', (accounts: string[]) => {
...
+              )
+
+              // ratio between the bucket's shouldn't change
+              assertEqualBN(
+                new BigNumber(mintableStable.dividedBy(tradeableGold)),
+                new BigNumber('4000')
+              )
+            })
+          })
+
```
Comment (nambrot) [inline packages/protocol/test/stability/exchange.ts:758]: Please add tests that check we are reverting wrong inputs

File: packages/protocol/test/stability/exchange.ts (hunk 5 above)
Comment (martinvol) [reply packages/protocol/test/stability/exchange.ts]: yeah, it's just to make the test run faster, will remove

File: packages/protocol/contracts/stability/Exchange.sol (hunk 1 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/Exchange.sol:4]: Yeap, I do need it for `Max` and `Min` functions

File: packages/protocol/contracts/stability/Exchange.sol (hunk 12 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/Exchange.sol]: In that case for a supply of zero to 1M, you get a cap of `1M*0.1`, then from there you get a cap of `Supply*0.1`, that it's bigger or equal to `1M*0.1`. Or at least that was my intention.

File: packages/protocol/test/stability/exchange.ts (hunk 14 above)
Comment (martinvol) [reply packages/protocol/test/stability/exchange.ts:758]: 👍

File: packages/protocol/contracts/stability/Exchange.sol (hunk 12 above)
Comment (nambrot) [reply packages/protocol/contracts/stability/Exchange.sol]: You are right, I misread the code. The comment led me to believe that the cap doesnt apply until a supply of 1M, so maybe I would reword the comment

**Comment by martinvol** [comment]
Will fix lint-checks, please make sure to review meanwhile!

File: packages/protocol/test/stability/exchange.ts [hunk 15]
```diff
@@ -626,6 +705,57 @@ contract('Exchange', (accounts: string[]) => {
...
+            })
+
+            it('has the correct getStableBucketCap', async () => {
+              assertEqualBN(
+                await exchange.getStableBucketCap(),
+                new BigNumber('2772727272727272816000000')
+              )
+            })
+
+            it('has the right price after the cap', async () => {
```
Comment (nambrot) [inline packages/protocol/test/stability/exchange.ts:744]: I would probably make this a another test group and set the rate in the `before` of the group and make separate assertions in separate tests. Right now it looks like you are just checking that the price/ratio of the buckets?

File: packages/protocol/test/stability/exchange.ts [hunk 16]
```diff
@@ -626,6 +705,57 @@ contract('Exchange', (accounts: string[]) => {
...
+              await mockSortedOracles.setMedianRate(
+                stableToken.address,
+                new BigNumber('0x20000000000000000').multipliedBy(2000)
+              ) // bump the price by 2000, leaving with 4K CELO per dollar
+
+              const [tradeableGold, mintableStable] = await exchange.getBuyAndSellBuckets(false)
```
Comment (nambrot) [inline packages/protocol/test/stability/exchange.ts:750]: What does mintableStable stand for?

Comment (martinvol) [reply packages/protocol/test/stability/exchange.ts:750]: Refers to the amount of stable that can be minted, it's also used [here](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/test/stability/exchange.ts#L515)

File: packages/protocol/contracts/stability/Exchange.sol [hunk 17]
```diff
@@ -34,8 +35,10 @@ contract Exchange is
   event MinimumReportsSet(uint256 minimumReports);
   event StableTokenSet(address indexed stable);
   event SpreadSet(uint256 spread);
+  event StableBucketFractionCapSet(uint256 stableBucketFractionCap);
```
Comment (yorhodes) [inline packages/protocol/contracts/stability/Exchange.sol:38]: would prefer if `Fraction` was the last suffix for decoding purposes
ie `stableBucketMaxFraction`

File: packages/protocol/contracts/stability/Exchange.sol [hunk 18]
```diff
@@ -54,6 +57,9 @@ contract Exchange is
   uint256 public updateFrequency;
   uint256 public minimumReports;
 
+  uint256 public minSupplyForStableBucketCap;
+  FixidityLib.Fraction public stableBucketFractionCap;
```
Comment (yorhodes) [inline packages/protocol/contracts/stability/Exchange.sol:61]: ditto

File: packages/protocol/contracts/stability/Exchange.sol [hunk 19]
```diff
@@ -368,9 +406,33 @@ contract Exchange is
...
+    if (cappedUpdatedStableBucket < updatedStableBucket) {
+      // resize Celo bucket
+      uint256 cappedUpdatedGoldBucket = exchangeRateDenominator.mul((maxStableBucketSize)).div(
+        exchangeRateNumerator
+      );
+      return (cappedUpdatedGoldBucket, cappedUpdatedStableBucket);
+
+    }
+
     return (updatedGoldBucket, updatedStableBucket);
```
Comment (yorhodes) [inline packages/protocol/contracts/stability/Exchange.sol]:
```suggestion
    // check if the bucket is bigger
    if (updatedStableBucket > maxStableBucketSize) {
      // resize Celo bucket
      uint256 cappedUpdatedGoldBucket = exchangeRateDenominator.mul((maxStableBucketSize)).div(
        exchangeRateNumerator
      );
      return (cappedUpdatedGoldBucket, maxStableBucketSize);
    }

    return (updatedGoldBucket, updatedStableBucket);
```

mixed references to CELO and gold are also slightly confusing

Comment (martinvol) [reply packages/protocol/contracts/stability/Exchange.sol]: yeah, I wanted to have no new references of Gold in the codebase.
